polygeist/lib/Dialect/Polygeist/Transforms/LoopInternalization.cpp

```diff
@@ -114,6 +102,97 @@ bool isCandidateLoopNest(LoopLikeOpInterface loop) {
   return true;
 }
 
+/// Get the size of unused shared local memory arena in bytes.
+unsigned getLocalMemoryRemain(gpu::GPUModuleOp &module,
+                              unsigned localMemorySize) {
+  assert(module.hasTrait<OpTrait::SymbolTable>() &&
+         "Expecting module with SymbolTable trait");
+  unsigned localMemoryRemain = localMemorySize;
+  module.walk([&localMemoryRemain](memref::GlobalOp global) {
+    MemRefType memRefTy = global.getType();
+    if (!isLocalAccessAddrSpace(memRefTy))
+      return WalkResult::advance();
+    unsigned globalSize =
+        memRefTy.getElementTypeBitWidth() * memRefTy.getNumElements() / 8;
+    if (globalSize >= localMemoryRemain) {
+      localMemoryRemain = 0;
+      return WalkResult::interrupt();
+    }
+    localMemoryRemain -= globalSize;
+    return WalkResult::advance();
+  });
+  return localMemoryRemain;
+}
+
+/// Get the require local memory for \p accTy.
+unsigned getReqdLocalMemory(sycl::AccessorType accTy,
+                            sycl::ReqdWorkGroupSize reqdWGSize) {
+  assert(!reqdWGSize.empty() && "Expecting non-empty reqdWGSize");
+  unsigned elemSize = accTy.getType().getIntOrFloatBitWidth();
+  unsigned memrefReqdLocalMemory = elemSize;
+  for (unsigned dim = 0; dim < accTy.getDimension(); ++dim) {
+    memrefReqdLocalMemory *= reqdWGSize[dim];
+  }
+  return memrefReqdLocalMemory;
+}
+
+/// Get the require local memory for memrefs in \p loopToSharedMemref.
+Optional<unsigned> getReqdLocalMemory(
+    DenseMap<LoopLikeOpInterface, std::set<Operation *>> &loopToSharedMemref,
+    sycl::ReqdWorkGroupSize reqdWGSize) {
+  if (reqdWGSize.empty())
+    return std::nullopt;
+
+  unsigned reqdLocalMemory = 0;
+  for (auto &entry : loopToSharedMemref) {
+    unsigned loopReqdLocalMemory = 0;
+    for (Operation *memref : entry.second) {
+      sycl::AccessorType accTy =
+          getAccessorType(cast<sycl::SYCLAccessorSubscriptOp>(memref));
+      loopReqdLocalMemory += getReqdLocalMemory(accTy, reqdWGSize);
+    }
+    // Memref in one loop can reuse local memory allocated for another loop.
+    reqdLocalMemory = std::max(reqdLocalMemory, loopReqdLocalMemory);
+  }
+  return reqdLocalMemory;
+}
+
+/// Create a GlobalOp for workgroup local memory.
+memref::GlobalOp getWGLocalMem(Operation *module, unsigned size) {
```

Comment:
Can we change Operation to gpu::GPUModuleOp

Reply:
done.